### PR TITLE
libc: Prefer to implement memfd on top of shm 

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary

since shm can work in protected and kernel mode too

## Impact

memfd

## Testing

internal

### CODE REVIEW

File: libs/libc/misc/lib_memfd.c [hunk 1]
```diff
@@ -45,7 +45,9 @@
 
 int memfd_create(FAR const char *name, unsigned int flags)
 {
-#ifdef CONFIG_FS_TMPFS
+#if defined(CONFIG_FS_SHMFS)
+  return shm_open(name, O_RDWR | flags, 0660);
+#elif defined(CONFIG_FS_TMPFS)
```
[the remark this answers is not in the document]
Comment (acassis) [reply libs/libc/misc/lib_memfd.c:50]: ping @xiaoxiang781216

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply libs/libc/misc/lib_memfd.c]: Done, @acassis

**Review by pkarashchenko — APPROVED** [review]
LGTM!

File: libs/libc/misc/lib_memfd.c [hunk 2]
```diff
@@ -32,20 +33,28 @@
...
 
-  snprintf(path, sizeof(path), LIBC_MEM_FD_VFS_PATH, name);
-  return open(path, O_RDWR | flags);
+  snprintf(path, sizeof(path), LIBC_MEM_FD_VFS_PATH_FMT, name);
+#if defined(CONFIG_LIBC_MEMFD_ON_SHMFS)
+  return shm_open(path, O_RDWR | flags, 0660);
+#elif defined(CONFIG_LIBC_MEMFD_ON_TMPFS)
+  mkdir(LIBC_MEM_LIBC_FD_VFS_PATH, 0666);
+  return open(path, O_RDWR | flags, 0660);
 #else
```
Comment (pkarashchenko) [inline libs/libc/misc/lib_memfd.c:58]:
```suggestion
#if defined(CONFIG_LIBC_MEMFD_ON_SHMFS) ||
    defined(CONFIG_LIBC_MEMFD_ON_TMPFS)
  char path[PATH_MAX];
    
  snprintf(path, sizeof(path), LIBC_MEM_FD_VFS_PATH_FMT, name);
#  if defined(CONFIG_LIBC_MEMFD_ON_SHMFS)
  return shm_open(path, O_RDWR | flags, 0660);
#  else
  mkdir(LIBC_MEM_LIBC_FD_VFS_PATH, 0666);
  return open(path, O_RDWR | flags, 0660);
#  endif
#else
  set_errno(ENOSYS);
  return ERROR;
#endif
```

Comment (xiaoxiang781216) [reply libs/libc/misc/lib_memfd.c]: Done.

File: libs/libc/misc/lib_memfd.c [hunk 3]
```diff
@@ -32,20 +33,30 @@
...
+#define LIBC_MEM_FD_VFS_PATH_FMT LIBC_MEM_FD_VFS_PATH "/%s"
 
 /****************************************************************************
  * Public Functions
  ****************************************************************************/
 
 int memfd_create(FAR const char *name, unsigned int flags)
 {
-#ifdef CONFIG_FS_TMPFS
+#if defined(CONFIG_LIBC_MEMFD_ON_SHMFS) || (CONFIG_LIBC_MEMFD_ON_TMPFS)
```
Comment (pkarashchenko) [inline libs/libc/misc/lib_memfd.c:50]:
```suggestion
#if defined(CONFIG_LIBC_MEMFD_ON_SHMFS) || defined(CONFIG_LIBC_MEMFD_ON_TMPFS)
```

File: libs/libc/misc/lib_memfd.c [hunk 4]
```diff
@@ -32,20 +33,30 @@
...
   char path[PATH_MAX];
 
-  snprintf(path, sizeof(path), LIBC_MEM_FD_VFS_PATH, name);
-  return open(path, O_RDWR | flags);
+  snprintf(path, sizeof(path), LIBC_MEM_FD_VFS_PATH_FMT, name);
+#  if defined(CONFIG_LIBC_MEMFD_ON_SHMFS)
+  return shm_open(path, O_RDWR | flags, 0660);
+#  elif defined(CONFIG_LIBC_MEMFD_ON_TMPFS)
```
Comment (pkarashchenko) [inline libs/libc/misc/lib_memfd.c:56]:
```suggestion
#  else
```

Comment (xiaoxiang781216) [reply libs/libc/misc/lib_memfd.c]: Done.

File: libs/libc/misc/lib_memfd.c (hunk 3 above)
Comment (xiaoxiang781216) [reply libs/libc/misc/lib_memfd.c]: I change the implementation, please review again, thanks.

File: libs/libc/misc/Kconfig [hunk 5]
```diff
@@ -81,9 +81,27 @@ config LIBC_FTOK_VFS_PATH
 	---help---
 		The relative path to where ftok will exist in the root namespace.
 
+choice
+	prompt "Select memfd implementation"
```
Comment (pkarashchenko) [inline libs/libc/misc/Kconfig:85]:
```suggestion
	prompt "Select memfd implementation"
	default LIBC_MEMFD_ERROR
```
? Or what is the default? `LIBC_MEMFD_SHMFS` if `FS_SHMFS`?

Comment (xiaoxiang781216) [reply libs/libc/misc/Kconfig:85]: I don't add the default value intentionally, kconfig tool select the first available choice in this case:

1. If user enable FS_SHMFS, memfd implement on top of shmfs
2. if user disable FS_SHMFS, but enable FS_TMPFS memfd impleemnt on top of tmpfs
3. if user disable both, memfd select LIBC_MEMFD_ERROR

This behavior is most user expect.

Comment (pkarashchenko) [reply libs/libc/misc/Kconfig:85]: Ok

**Review by pkarashchenko — APPROVED** [review]

File: libs/libc/misc/Kconfig [hunk 6]
```diff
@@ -81,9 +81,27 @@ config LIBC_FTOK_VFS_PATH
...
+
 config LIBC_MEM_FD_VFS_PATH
 	string "Relative path to memfd storage"
 	default "memfd"
-	depends on FS_TMPFS
+	depends on !LIBC_MEMFD_ERROR
 	---help---
 		The relative path to where memfd will exist in the tmpfs namespace.
+
+endif
```
Comment (pkarashchenko) [inline libs/libc/misc/Kconfig:107]: Why `endif` is needed here?

Comment (xiaoxiang781216) [reply libs/libc/misc/Kconfig]: My fault, fixed, thanks.

**Comment by xiaoxiang781216** [comment]
ping @pkarashchenko

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**